qutip/solver/flimesolve.py

```diff
@@ -59,107 +59,103 @@ def _floquet_rate_matrix(floquet_basis,
-                                  c_op_Fourier_amplitudes_list, np.conj(c_op_Fourier_amplitudes_list))
-        rate_products_idx = np.argwhere(abs(rate_products) != 0)
-
-        args_key = list(omega.keys())[0]
-        valid_indices = [tuple(indices) for indices in rate_products_idx
-                         if delta(*tuple(indices)) == 0
-                         or abs((rate_products[tuple(indices)]
-                                 / (omega[args_key]
-                                    * delta(*tuple(indices))))**(-1)) <= time_sense
-                         and abs(omega[args_key] * delta(*tuple(indices))) < Nt / 2]
-
-        delta_dict = {}
-        for indices in valid_indices:
-            try:
-                delta_dict[delta(*indices)].append(indices)
-            except KeyError:
-                delta_dict[delta(*indices)] = [indices]
-
-        for key in delta_dict.keys():
-            mask_array = np.zeros((Hdim, Hdim, Hdim, Hdim, Nt, Nt))
-            for indices in delta_dict[key]:
-                mask_array[indices] = True
-            valid_c_op_products = rate_products * mask_array
-
-            # using c = ap,d = bp,k=lp
-            flime_FirstTerm = c_op_rates[cdx] \
-                * np.einsum('abcdlk,ma,nc,pb,qd->mnpq',
-                            valid_c_op_products,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-
-            flime_SecondTerm = c_op_rates[cdx] \
-                * np.einsum('abcdlk,ac,md,pb,qn->mnpq',
-                            valid_c_op_products,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-
-            flime_ThirdTerm = c_op_rates[cdx] \
-                * np.einsum('abcdlk,ac,nb,pm,qd->mnpq',
-                            valid_c_op_products,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-            try:
-                total_R_tensor[key] += np.reshape(flime_FirstTerm - (1 / 2)
-                                                  * (flime_SecondTerm +
-                                                     flime_ThirdTerm),
-                                                  (Hdim**2, Hdim**2))
-            except KeyError:
-                total_R_tensor[key] = np.reshape(flime_FirstTerm - (1 / 2)
-                                                 * (flime_SecondTerm +
-                                                    flime_ThirdTerm),
-                                                 (Hdim**2, Hdim**2))
+
+        for l, k in product(np.arange(Nt), repeat=2):
+            delta_shift = delta_m + (l - k)
```

Comment:
We want small, `delta_shift`, when `l == k` there are evident cancellation, but otherwise there are usually no included values. By looping over them, we can skip most case where `l != k` and use smaller matrices, saving both computation and memory.

Reply:
This makes sense. I think if I were to fully implement this, I could rewrite a lot of the rate matrix construction to be a bit faster. I would need a little time to fully flesh it out, but I could probably do something like:

1. Find the largest difference of (l-k) that any given 'time_sense' argument would support (this would be the most difficult step because of the more qualitative implementation of 'time_sense' I currently have built into FLiMESolve).
2. Use this to constrict the sums over l,k
3. Use the entirety of the resulting array to build FLiMESolve instead of checking every element for the condition imposed by 'time_sense'. 

I'll work on this today and see what I can't figure out.

Reply:
Oh! It looks like you've already done all the work for that. Wow, thank you!